repos/system_upgrade/el8toel9/actors/checkifcfg/libraries/ifcfg.py

```diff
@@ -0,0 +1,145 @@
+                    got_type = 'bond'
+                elif got_type != 'bond':
+                    bad_type = True
+
+            if key == 'NM_CONTROLLED' and value.lower() not in TRUE_VALUES:
+                nm_controlled = False
+
+        if got_type not in GOOD_TYPES:
+            bad_type = True
+
+        # Don't bother reporting the file for NM_CONTROLLED=no
+        # if its type is not supportable with NetworkManager anyway
+        if bad_type is True:
+            bad_type_files.append(f.path)
+        elif nm_controlled is False:
+            not_controlled_files.append(f.path)
+
+    if rule_files:
+        title = 'Rule file detected'
+        summary = ('RHEL 9 does not support the legacy network-scripts'
+                   ' package that was deprecated in RHEL 8 in favor of'
+                   ' NetworkManager. Rule files present in the system will'
+                   ' not work with NetworkManager.')
+        remediation = 'Replace rule files with a NetworkManager dispatcher script.'
+        reporting.create_report([
+            reporting.Title(title),
+            reporting.Summary(summary),
+            reporting.Remediation(hint=remediation),
+            reporting.Severity(reporting.Severity.HIGH),
+            reporting.Tags([reporting.Tags.NETWORK, reporting.Tags.SERVICES]),
+            reporting.Flags([reporting.Flags.INHIBITOR]),
+            reporting.RelatedResource('package', 'NetworkManager'),
+            reporting.ExternalLink(
+                title='NetworkManager-dispatcher - Dispatch user scripts for NetworkManager',
+                url='https://networkmanager.dev/docs/api/latest/NetworkManager-dispatcher.html',
+            ),
+        ] + [
+            reporting.RelatedResource('file', fname)
+            for fname in rule_files
+        ])
+
+    if bad_type_files:
+        title = 'Network configuration for unsupported device types detected'
+        summary = ('RHEL 9 does not support the legacy network-scripts'
+                   ' package that was deprecated in RHEL 8 in favor of'
+                   ' NetworkManager. Files for device types that are not'
+                   ' supported by NetworkManager are present in the system.')
+        remediation = ('Consult the nm-settings-ifcfg-rh(5) manual for'
+                       ' valid types of ifcfg files. Remove configuration'
+                       ' files that can not be supported.')
+        reporting.create_report([
+            reporting.Title(title),
+            reporting.Summary(summary),
+            reporting.Remediation(hint=remediation),
+            reporting.Severity(reporting.Severity.HIGH),
+            reporting.Tags([reporting.Tags.NETWORK, reporting.Tags.SERVICES]),
+            reporting.Flags([reporting.Flags.INHIBITOR]),
+            reporting.RelatedResource('package', 'NetworkManager'),
+        ] + [
+            reporting.RelatedResource('file', fname)
+            for fname in bad_type_files
```

Comment:
How about including the external link for `nm-settings-ifcfg-rh(5) manual` here?

Reply:
I intentionally omitted it because if there's some unsupported device type, then we can't guess what type it actually is or what is it supposed to do, and the ifcfg manual is unlikely to help.

It's difficult to guess what could be happening at that time, but it's likely that the only option the user has is to remove the file and do whatever they were trying to achieve differently (e.g. with a shell script or something).